### PR TITLE
PERF-3476 add workload for $group queries that spill to disk

### PR DESCRIPTION
**Pull request by dstorch**
Proposed detailed commit message:

> Also extends the 'CrudActor' to support aggregate operations.

Hopefully reviewers are ok with lumping in the changes to the `CrudActor` here. The lack of support for "aggregate" in the `CrudActor` has led to the use of explain in genny tests in the past (e.g. for TPC-H; see [PERF-3251](https://jira.mongodb.org/browse/PERF-3251)), so it seemed useful to provide this functionality to support future agg perf tests.

I kicked off a sys-perf patch build here: https://spruce.mongodb.com/version/635ad15c5623435c5fb6d9a3/tasks

And a genny patch build here: https://spruce.mongodb.com/version/635ad2360305b91ce272e733/tasks

### CODE REVIEW

**Comment by dalyd** [comment]
Could you split this PR into 2? One for the CrudActor change, and one for the workload? I suspect there will be some additions for the CrudActor change, and it would be nice to separate the functionality from the use of the functionality.

**Comment by dalyd** [comment]
Could you add an example to https://github.com/mongodb/genny/blob/master/src/workloads/docs/CrudActor.yml or to a new file in docs? 

Also could you add some test cases to https://github.com/mongodb/genny/blob/master/src/cast_core/test/CrudActorYamlTests.yml or as appropriate for the new functionality?

**Comment by dstorch** [comment]
I'm splitting this PR into two per @dalyd's suggestion. This PR will be responsible only for adding the new $group spilling workload. It will become dependent on a yet-to-be created PR that extends the `CrudActor`. I filed [PERF-3485](https://jira.mongodb.org/browse/PERF-3485) for the `CrudActor` changes and will post a PR soon. David's other comments will be addressed in the new PR, not this one.

Given the split, I think that Mihai can be the sole reviewer of this change. I'll ask again in #performance-tooling-users in order to find a reviewer for the `CrudActor` change. @mtandrei please take a look!

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+GlobalDefaults:
+  Database: &Database test
+  Collection: &Collection Collection0
+  DocumentCount: &DocumentCount 2_000_000
+  FirstFieldUniqueValCount: &FirstFieldUniqueValCount 20
+  RandomStrLength: &RandomStrLength 60
+  MaxPhases: &MaxPhases 5
+
+Actors:
+# Phase 0: Insert documents into the collection.
```
Comment (mtandrei) [inline src/workloads/execution/GroupSpillToDisk.yml:23]: [optional] recommend prepending a phase that drops the collection; this is useful for local testing: https://github.com/mongodb/genny/blob/58901acf3ed1a4dd41f55f41f1d46689ff804e51/src/workloads/execution/LookupSBEPushdown.yml#L234

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          OperationCommand:
+            Pipeline: [{$group: {_id: "$secondField", count: {$sum: 1}}}]
+            Options:
+              AllowDiskUse: true
+
+AutoRun:
+- When:
+    mongodb_setup:
+      $eq:
+      # TODO PERF-3480: Enable this workload against 'standalone', 'standalone-all-feature-flags'
```
Comment (mtandrei) [inline src/workloads/execution/GroupSpillToDisk.yml:116]: Is the thinking here that SBE Perf is so bad currently that we risk timing out? Unless there's a risk of the workload not finishing, we should enable this on the SBE variants to get a baseline measurement.

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      # and 'standalone-sbe' once the SBE performance improves. These queries will use SBE unless
+      # classic is forced, but they run for too long at the moment due to the performance problem
+      # reported in SERVER-70395.
+      - standalone-classic-query-engine
```
Comment (mtandrei) [inline src/workloads/execution/GroupSpillToDisk.yml:120]: Could you reconfigure your patch build to run 'auto run' on the standalone classic variant: https://spruce.mongodb.com/version/635ad15c5623435c5fb6d9a3/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC ?

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        - OperationName: aggregate
+          OperationCommand:
+            Pipeline: [{$group: {_id: "$firstField", list: {$push: "$secondField"}}}]
+            Options:
+              AllowDiskUse: true
+
+# This query groups by "secondField", which has many distinct values. Spilling is required since
+# there are many groups, but the state for each individual group is small.
+- Name: SumAccumulatorSpillDueToManyGroups
+  Type: CrudActor
```
Comment (mtandrei) [inline src/workloads/execution/GroupSpillToDisk.yml:95]: [question] what is the rationale for supporting aggregates in the CrudActor (i.e. what advantage does this provide [to running aggregates like any other command](https://github.com/mongodb/genny/blob/58901acf3ed1a4dd41f55f41f1d46689ff804e51/src/phases/execution/LookupCommands.yml#L72-L85)) In the description you mention:
`The lack of support for "aggregate" in the CrudActor has led to the use of explain in genny tests in the past`
But it seems that you can get away with not using explain?

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        Database: *Database
+        MultipleThreadsPerCollection: true
+        CollectionCount: 1
+        DocumentCount: *DocumentCount
+        BatchSize: 1000
+        Document:
+          firstField: {^RandomInt: {min: 1, max: *FirstFieldUniqueValCount}}
+          secondField: {^RandomString: {length: *RandomStrLength, alphabet: "0123456789"}}
+
+- Name: Quiesce
```
Comment (mtandrei) [inline src/workloads/execution/GroupSpillToDisk.yml:42]: [nit] comment why we need to quiesce

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  Phases:
+    OnlyActiveInPhases:
+      Active: [1]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: 1
+        Threads: 1
+
+# This query has a small number of groups, but each group is a very large set.
+- Name: AddToSetAccumulatorWithLargeSets
```
Comment (mtandrei) [inline src/workloads/execution/GroupSpillToDisk.yml:55]: [mega nit] prepend 'Group' to these tests ($addToSet, $push, $sum are also supported in $setWindowsFields)

**Review by mtandrei — APPROVED** [review]
LGTM with some minor nits and questions; thanks Dave!

File: src/workloads/execution/GroupSpillToDisk.yml (hunk 3 above)
Comment (dstorch) [reply src/workloads/execution/GroupSpillToDisk.yml:136]: D'oh! My bad, thanks for catching this. Done, hopefully it passes!

**Review by dstorch — COMMENTED** [review]
@mtandrei I pushed changes addressing your comments, and I'm publishing replies to your comment threads. Please take another look -- your LGTM got cleared away (automatically, I think?).

Also, note that I changed the base branch of this PR to `PERF-3485` now that I have the `CrudActor` changes up for review: https://github.com/mongodb/genny/pull/761. I've never tried to change the base branch of a PR mid-flight before, so hopefully it doesn't mess up comment threads and/or the diff!

**Review by mtandrei — APPROVED** [review]
Still LGTM

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  Phases:
+    OnlyActiveInPhases:
+      Active: [0]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: 1
+        Database: *Database
+        Operations:
+        - OperationName: RunCommand
+          OperationCommand: {dropDatabase: 1}
```
[the remark this answers is not in the document]
Comment (dstorch) [reply src/workloads/execution/GroupSpillToDisk.yml:37]: Yeah, my understanding is that each workload can assume it is the sole entity running against the cluster. I don't think we currently have a notion of simulating complicated workloads by running multiple genny workloads in parallel. So yes, I believe the "dropDatabase" command is ok here. I also copied this logic from the $lookup SBE pushdown workloads, so we're doing the same elsewhere already.

File: src/workloads/execution/GroupSpillToDisk.yml [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+- group
+- spill
+
+GlobalDefaults:
+  Database: &Database test
+  Collection: &Collection Collection0
+  DocumentCount: &DocumentCount 2_000_000
+  FirstFieldUniqueValCount: &FirstFieldUniqueValCount 20
+  RandomStrLength: &RandomStrLength 60
+  MaxPhases: &MaxPhases 5
```
[the remark this answers is not in the document]
Comment (dstorch) [reply src/workloads/execution/GroupSpillToDisk.yml:20]: Yikes, good catch, done! The workload seemed to do the right thing without this, but maybe it was actually doing the wrong thing? Not sure.

**Comment by dstorch** [comment]
Thanks @mtandrei! I've responded to your final remaining comments, but did not push any new code. When https://github.com/mongodb/genny/pull/761 lands I'll change the base branch to `master`, at which point I may require another approval to merge? Not sure, but I'll let you know!

**Comment by dstorch** [comment]
evergreen merge

**Pull request merged**